source/common/upstream/load_stats_reporter.cc

```diff
@@ -0,0 +1,139 @@
+                                     ClusterManager& cluster_manager, Stats::Scope& scope,
+                                     LoadStatsAsyncClientPtr async_client,
+                                     Event::Dispatcher& dispatcher)
+    : cm_(cluster_manager), stats_{ALL_LOAD_REPORTER_STATS(
+                                POOL_COUNTER_PREFIX(scope, "load_reporter."))},
+      async_client_(std::move(async_client)),
+      service_method_(*Protobuf::DescriptorPool::generated_pool()->FindMethodByName(
+          "envoy.api.v2.EndpointDiscoveryService.StreamLoadStats")) {
+  request_.mutable_node()->MergeFrom(node);
+  retry_timer_ = dispatcher.createTimer([this]() -> void { establishNewStream(); });
+  response_timer_ = dispatcher.createTimer([this]() -> void { sendLoadStatsRequest(); });
+  establishNewStream();
+}
+
+LoadStatsReporter::LoadStatsReporter(const envoy::api::v2::Node& node,
+                                     ClusterManager& cluster_manager, Stats::Scope& scope,
+                                     const std::string& remote_cluster_name,
+                                     Event::Dispatcher& dispatcher)
+    : LoadStatsReporter(
+          node, cluster_manager, scope,
+          LoadStatsAsyncClientPtr(new Grpc::AsyncClientImpl<envoy::api::v2::LoadStatsRequest,
+                                                            envoy::api::v2::LoadStatsResponse>(
+              cluster_manager, remote_cluster_name)),
+          dispatcher) {}
+
+void LoadStatsReporter::setRetryTimer() {
+  retry_timer_->enableTimer(std::chrono::milliseconds(RETRY_DELAY_MS));
+}
+
+void LoadStatsReporter::establishNewStream() {
+  ENVOY_LOG(debug, "Establishing new gRPC bidi stream for {}", service_method_.DebugString());
+  stream_ = async_client_->start(service_method_, *this);
+  if (stream_ == nullptr) {
+    ENVOY_LOG(warn, "Unable to establish new stream");
+    handleFailure();
+    return;
+  }
+
+  request_.mutable_cluster_stats()->Clear();
+  sendLoadStatsRequest();
+}
+
+void LoadStatsReporter::sendLoadStatsRequest() {
+  request_.mutable_cluster_stats()->Clear();
+  for (const std::string& cluster_name : clusters_) {
+    auto cluster_info_map = cm_.clusters();
+    auto it = cluster_info_map.find(cluster_name);
+    if (it == cluster_info_map.end()) {
+      ENVOY_LOG(debug, "Cluster {} does not exist", cluster_name);
+      continue;
+    }
+    auto& cluster = it->second.get();
+    auto* cluster_stats = request_.add_cluster_stats();
+    cluster_stats->set_cluster_name(cluster_name);
+    for (auto& hosts : cluster.hostsPerLocality()) {
+      auto* locality_stats = cluster_stats->add_upstream_locality_stats();
+      ASSERT(hosts.size() > 0);
+      locality_stats->mutable_locality()->MergeFrom(hosts[0]->locality());
+      uint64_t rq_success = 0;
+      uint64_t rq_error = 0;
+      uint64_t rq_active = 0;
+      for (auto host : hosts) {
+        rq_success += host->stats().rq_success_.latch();
```

Comment:
Per offline discussion do we have comments/docs anywhere on how we are hacking latching somewhat? It might be worth it to put a warning on the stat definitions themselves. I can't remember if we did that or not.

Reply:
Already comments in https://github.com/envoyproxy/envoy/blob/master/include/envoy/upstream/host_description.h#L19 and the changes to `upstream.h`, but I will be more explicit about latching.